File: php/examples/workflow_with_wasm_plugins.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use NativeBPM\Client\Builder\Workflow;
use NativeBPM\Client\Configuration;
use NativeBPM\Client\Api\DefaultApi;
use NativeBPM\Client\Model\StartInstanceRequest;

echo "=== NativeBPM PHP SDK: Workflow with Guest WASM Plugins ===\n";

// 1. Build workflow dynamically using Workflow as Code Fluent API
echo "🔨 Building workflow dynamically using Fluent API...\n";
$workflow = new Workflow("wasm-demo", "Workflow with Guest WASM Plugins");

// Chain starting from first service task
$workflow->start()
    ->service("calculate", "Calculate Totals", "payment_topic", ["wasmPath" => "./calculate_total.wasm"])
    ->ai("aiCheck", "AI Fraud Guard", [
        "provider" => "google",
        "model" => "gemini-2.5-flash",
        "prompt" => 'Analyze transaction for fraud: ${orderAmount}',
        "resultVar" => "isFraudulent"
    ])
    ->when(Workflow::V('isFraudulent')->eq(true))->then(function($b) {
        $b->user("userTask", "Manual Fraud Approval", ["assignee" => "security_officer"]);
    })
    ->else(function($b) {
        // empty branch (will auto-route to end event)
    });

// 2. Deploy and start process definition using the REST API client
$config = Configuration::getDefaultConfiguration()
    ->setHost("http://localhost:8080")
    ->setApiKey("Authorization", "Bearer test-bearer-token");

$api = new DefaultApi(new GuzzleHttp\Client(), $config);

echo "\nDeploying to NativeBPM engine (JSON AST compiled server-side)...\n";

try {
    // Deploy process definition directly via JSON AST
    $definition = $api->deployWorkflow($workflow);
    echo "✓ Deployed process definition (hash: " . $definition->getHash() . ")\n";

    // Start a process instance with input variables
    $startRequest = new StartInstanceRequest();
    $startRequest->setInstanceId(sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
        mt_rand(0, 0xffff), mt_rand(0, 0xffff),
        mt_rand(0, 0xffff),
        mt_rand(0, 0x0fff) | 0x4000,
        mt_rand(0, 0x3fff) | 0x8000,
        mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
    ));
    $startRequest->setBusinessKey("tx-8837");
    $startRequest->setVariables([
        "orderAmount" => 2500
    ]);

    $instance = $api->startInstance($definition->getId(), $startRequest);
    echo "✓ Started process instance ID: " . $instance->getId() . " (completed: " . ($instance->getCompleted() ? 'true' : 'false') . ")\n";

} catch (Exception $e) {
    echo "Note: Local API Engine deployment skipped (ensure local server is running on :8080). Details: " . $e->getMessage() . "\n";
}

File: python/examples/workflow_with_wasm_plugins.py
from nativebpm import Client, Workflow, v

def main():
    print("=== NativeBPM Python SDK: Workflow with Guest WASM Plugins ===")

    # 1. Build workflow as code using Fluent API method chaining
    workflow = Workflow('wasm-demo', 'Workflow with Guest WASM Plugins')
    
    # Chain starting from the start event
    workflow\
        .service('calculate', 'Calculate Totals', 'payment_topic', wasm='./calculate_total.wasm')\
        .ai('aiCheck', 'AI Fraud Guard',
            provider='google',
            model='gemini-2.5-flash',
            prompt='Analyze transaction for fraud: ${orderAmount}',
            result_var='isFraudulent')\
        .when(v('isFraudulent').eq(True)).then(lambda b: (
            b.user('userTask', 'Manual Fraud Approval', assignee='security_officer')
        ))\
        .Else(lambda b: (
            None
        ))
    
    # 2. Deploy and start process definition using the Fluent Client API
    client = Client("http://localhost:8080", "test-bearer-token")
    
    print("\nDeploying to NativeBPM engine (JSON AST compiled server-side)...")
    try:
        # Deploy process definition directly via Workflow object
        definition = client.deploy(workflow)
        print(f"✓ Deployed process definition (hash: {definition.hash})")
        
        # Start a process instance with input variables
        instance = client.instances().start("wasm-demo")\
            .with_business_key("tx-8837")\
            .with_variable("orderAmount", 2500)\
            .send()
        print(f"✓ Started process instance ID: {instance.id} (completed: {instance.completed})")
        
    except Exception as e:
        print(f"Note: Local API Engine deployment skipped (ensure local server is running on :8080). Details: {e}")

if __name__ == '__main__':
    main()
